src/deploy/functions/iambinder.ts

```diff
@@ -0,0 +1,298 @@
+      }
+      bindingInternal.add(role, members, condition);
+    }
+    return bindingInternal;
+  }
+}
+
+/**
+ * IamBinder generates necessary IAM binding given different types of function endpoints.
+ */
+export class IamBinder {
+  additions: Record<Resource, IamBindings>; // resource -> bindings
+  storageServiceAccount: string | undefined;
+
+  constructor(private readonly projectId: string, private readonly projectNumber: string) {
+    this.additions = {};
+  }
+
+  async apply(): Promise<iam.Policy[]> {
+    const updates = Object.entries(this.additions).map(async ([resource, bindings]) => {
+      return await IamBinder.updatePolicy(resource as Resource, bindings);
+    });
+    return Promise.all(updates);
+  }
+
+  async addEndpoints(endpoints: backend.Endpoint[]): Promise<void> {
+    for (const endpoint of endpoints) {
+      await this.addEndpoint(endpoint);
+    }
+  }
+
+  private async addEndpoint(endpoint: backend.Endpoint): Promise<void> {
+    let serviceAccount = this.defaultComputeServiceAccount(endpoint);
+    if (endpoint.serviceAccountEmail && endpoint.serviceAccountEmail !== "default") {
+      serviceAccount = `serviceAccount:${endpoint.serviceAccountEmail}`;
+    }
+
+    for (const secret of endpoint.secretEnvironmentVariables || []) {
+      this.addBinding(
+        `//secretmanager.googleapis.com/projects/${this.projectNumber}/secrets/${secret.secret}`,
+        "roles/secretmanager.secretAccessor",
+        [serviceAccount]
+      );
+    }
+
+    if (backend.isEventTriggered(endpoint)) {
+      if (endpoint.platform === "gcfv2") {
+        // Eventarc requires pubsub service account to have serviceAccountTokenCreator role.
+        // See https://cloud.google.com/eventarc/docs/roles-permissions#triggertypes-roles.
+        this.addBinding(
+          `//cloudresourcemanager.googleapis.com/projects/${this.projectNumber}`,
+          "roles/iam.serviceAccountTokenCreator",
+          [`serviceAccount:service-${this.projectNumber}@gcp-sa-pubsub.iam.gserviceaccount.com`]
+        );
+        // Service account associated w/ eventarc trigger must have permission to invoke a Cloud Run instance.
+        // We don't allow users to specify this service account hence it will always be the default compute service
+        // account. Usually, the default service account has EDITOR role which includes permission to invoke a Cloud
+        // Run instance, but given strict org policy it may not have this permission.
+        // See https://cloud.google.com/compute/docs/access/service-accounts#default_service_account.
+        this.addBinding(
+          `//cloudresourcemanager.googleapis.com/projects/${this.projectNumber}`,
+          "roles/run.invoker",
+          [this.defaultComputeServiceAccount(endpoint)]
+        );
+        if (v2Events.isStorageTriggered(endpoint)) {
+          // Storage account associated w/ Cloud Storage must have permission to publish pubsub messages.
+          this.addBinding(
```

Comment:
Huh I assumed that service abstraction is roughly "grouping endpoints by event types (into "service" grouping)" which I hoped to emulate here. Do you think I'm missing something here?

Reply:
I was just thinking out loud haha. I think I'm okay with doing that we're doing here since `storage` on v2 seems like the only special permissions case that we have. The `service` abstraction was originally created to avoid a bunch of `if` checks and instead just do a `noop` per @inlined 's direction

Reply:
Ah I see what you mean. I like the explicitness of bunch of if statements, but I can see it being unwieldy when there is TOO much of that 🤔 . I hope this list isn't getting that much bigger, but if this is more confusing and you think service abstraction makes more sense, I can bring that back here.

Reply:
I'm okay with keeping this as a one-off check since it seems like storage is a special case, we can revisit if other providers pop up in the future.